### ISSUE TITLE
Should Storage implement ops::Index?

### ISSUE THREAD

**Issue by leod** [slide-rs/specs#347]
As I understand it, the idiomatic way of accessing components is via a `Join`. However, sometimes this isn't easily possible. In my case, I have components that reference other entities by storing an `Entity`. For example, an enemy entity might want to store what entities it is currently targeting.

To access components of the referenced entity, I'm using lots of `component.get(referenced_entity).unwrap()`, which isn't very readable. Thus, I think it would be nice if `Storage` implemented `std::ops::Index`. Or are there better ways to go about this?

(As a side-node, I'm a bit uneasy with storing and passing around `Entity` for this purpose, since it basically introduces duck typing and you have to be careful that the entity has the correct components. If rust ever gets variadic templates, it might be nice to have a type `EntityHandle<T...>` which expresses that the entity has at least the components given in the template parameters.)

**Comment by WaDelma** [comment]
There is new api (thats not published to crates yet) for getting multiple components for certain entity which makes that cleaner:
```rust
let (left_shape, left_pose, left_next_pose) =
    (&shapes, &poses, &next_poses).join().get(left_entity).unwrap();
```

Implementing `std::ops::Index` would remove need for `unwrap`ing, but I don't know if I really like that. I think the default indexing not returning `Option` is kinda misfeature.

Also that `EntityHandle` idea wouldn't work: You can always remove components. Also it would still require you to check that `Entity` is alive.

**Comment by leod** [comment]
The new api looks good enough, since it means I'm only calling `unwrap` once.

Agreed about your other points.

**Comment by torkleyy** [comment]
You should not call unwrap because it will panic if the component or entity was removed.

**Comment by leod** [comment]
Of course, I'm operating under the assumption that I know that the entity still lives and want to access several of its components.

**Comment by torkleyy** [comment]
Okay :)

**Comment by omni-viral** [comment]
@leod such assumptions get broken on regular basis once the project is big enough. You should be careful with them :smile:

**Comment by leod** [comment]
@omni-viral Agreed, which is why I said that I'm uneasy with storing or passing around `Entity`. But I think that unwrapping is unavoidable in certain situations . Some examples that I have encountered:

1. Physically joining two entities, by storing `Entity` either in a component or a resource (in pairs). This requires the entities to have a `Position` component. A joint between position-less entities is simply a bug, so unwrapping makes sense.

2. Interaction handlers that are called when two entities intersect. In my multiplayer game, every entity belongs to some entity class. An entity class specifies, among other things, the components that should always be present for entities of that class (this helps in delta encoding of game state snapshots). For example, `"player"` entities should *always* have `Position`, `Velocity`, and `Player`. Now, I can add an interaction handler like this:
```rust
interaction::add("player", "enemy", handler);
fn handler(world: &World, player: Entity, enemy: Entity) {
    world.write::<Position>().get_mut(player).unwrap().0 = ...;
}
```
`"player"` always has `Position` by definition, so it makes sense to unwrap.

3. An entity that "owns" and manages/coordinates multiple sub-entities. Here, again, I know exactly what class the sub-entities belong to, but I have to unwrap to access their components.

So basically, things seem to get hairy whenever I want to access two entities *at the same time*. I'd definitely be interested to know if anyone has a better, less unwrappy, approach to this!

**Comment by omni-viral** [comment]
@leod don't forget that entity may be destroyed when you query it's component. In this case unwrapping is not alowed.
